File: src/test/java/io/github/aalsanie/codes/ValidationResultTest.java
package io.github.aalsanie.codes;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.Test;

class ValidationResultTest {
    @Test
    void validIsSingletonAndConvertsToOk() {
        ValidationResult valid = ValidationResult.valid();
        assertSame(valid, ValidationResult.valid());
        assertTrue(valid.isValid());
        assertFalse(valid.isInvalid());
        assertTrue(valid.issues().isEmpty());
        assertEquals("Valid", valid.toString());
        assertEquals(StandardOutcomes.OK, valid.toOutcome(StandardOutcomes.INVALID_ARGUMENT).getDefinition());
    }

    @Test
    void invalidCopiesIssuesAndConvertsToExplicitFailure() {
        ArrayList<Issue> issues = new ArrayList<>();
        issues.add(Issue.at("email", "Invalid."));
        ValidationResult result = ValidationResult.invalid(issues);
        issues.clear();

        assertTrue(result.isInvalid());
        assertEquals(1, result.issues().size());
        assertEquals(result, ValidationResult.invalid(Issue.at("email", "Invalid.")));
        assertEquals(result.hashCode(), ValidationResult.invalid(Issue.at("email", "Invalid.")).hashCode());
        assertTrue(result.toString().contains("Invalid"));

        Outcome outcome = result.toOutcome(StandardOutcomes.INVALID_ARGUMENT, "request validation failed");
        assertEquals(StandardOutcomes.INVALID_ARGUMENT, outcome.getDefinition());
        assertEquals("request validation failed", outcome.getDetail());
        assertEquals(result.issues(), outcome.getIssues());
    }

    @Test
    void combinesIndependentResults() {
        ValidationResult result = ValidationResult.combine(
            ValidationResult.valid(),
            ValidationResult.invalid(Issue.at("email", "Invalid.")),
            ValidationResult.invalid(Issue.at("name", "Required."))
        );
        assertEquals(2, result.issues().size());
        assertTrue(ValidationResult.combine(List.of()).isValid());
        assertThrows(NullPointerException.class, () -> ValidationResult.combine((ValidationResult[]) null));
        assertThrows(NullPointerException.class, () -> ValidationResult.combine((List<ValidationResult>) null));
        assertThrows(NullPointerException.class, () -> ValidationResult.combine(List.of(ValidationResult.valid(), null)));
    }

    @Test
    void rejectsInvalidPoliciesAndEmptyInvalidResults() {
        OutcomeDefinition success = OutcomeDefinition.custom(
            "com.example", "CREATED", OutcomeState.SUCCEEDED, "Created."
        );
        assertThrows(IllegalArgumentException.class, () -> ValidationResult.valid().toOutcome(success));
        assertThrows(IllegalArgumentException.class, () -> ValidationResult.invalid(List.of()));
        assertThrows(NullPointerException.class, () -> ValidationResult.invalid((Issue) null));
        assertThrows(NullPointerException.class, () -> ValidationResult.invalid((List<Issue>) null));
        assertThrows(NullPointerException.class, () -> ValidationResult.valid().toOutcome(null));
    }
}

File: src/test/kotlin/io/github/aalsanie/codes/ValidationResultTest.kt
package io.github.aalsanie.codes

import org.junit.jupiter.api.Test

class ValidationResultTest {
    @Test fun validHasNoIssues() {
        val result = ValidationResult.valid()
        assertTrue(result.isValid)
        assertFalse(result.isInvalid)
        assertTrue(result.issues().isEmpty())
    }

    @Test fun invalidContainsIssue() {
        val issue = Issue.of("Bad.")
        val result = ValidationResult.invalid(issue)
        assertFalse(result.isValid)
        assertTrue(result.isInvalid)
        assertEquals(listOf(issue), result.issues())
    }

    @Test fun invalidRejectsEmptyIssues() {
        assertFails<IllegalArgumentException> { ValidationResult.invalid(emptyList()) }
    }

    @Test fun invalidDefensivelyCopiesIssues() {
        val source = mutableListOf(Issue.of("Bad."))
        val result = ValidationResult.invalid(source)
        source.clear()
        assertEquals(1, result.issues().size)
    }

    @Test fun combinesValidResults() {
        assertSame(ValidationResult.Valid, ValidationResult.combine(ValidationResult.valid(), ValidationResult.valid()))
    }

    @Test fun combinesAndPreservesValidationIssues() {
        val first = Issue.at("a", "Bad a.")
        val second = Issue.at("b", "Bad b.")
        val result =
            ValidationResult.combine(
                ValidationResult.invalid(first),
                ValidationResult.valid(),
                ValidationResult.invalid(second),
            )
        assertEquals(listOf(first, second), result.issues())
    }

    @Test fun combinesListOverload() {
        val issue = Issue.of("Bad.")
        assertEquals(listOf(issue), ValidationResult.combine(listOf(ValidationResult.invalid(issue))).issues())
    }

    @Test fun validConvertsToOkWithExplicitFailureDefinition() {
        val outcome = ValidationResult.valid().toOutcome(StandardOutcomes.INVALID_ARGUMENT)
        assertSame(StandardOutcomes.OK, outcome.definition)
    }

    @Test fun invalidConvertsToExplicitFailureDefinition() {
        val issue = Issue.of("Bad.")
        val outcome =
            ValidationResult
                .invalid(issue)
                .toOutcome(StandardOutcomes.FAILED_PRECONDITION, "validation rejected")
        assertSame(StandardOutcomes.FAILED_PRECONDITION, outcome.definition)
        assertEquals(listOf(issue), outcome.issues)
        assertEquals("validation rejected", outcome.detail)
    }

    @Test fun invalidCanUseCustomFailureDefinition() {
        val definition =
            OutcomeDefinition.custom(
                "com.example.validation",
                "ORDER_NOT_MODIFIABLE",
                OutcomeState.FAILED,
                "The order cannot be modified.",
            )
        val outcome = ValidationResult.invalid(Issue.of("Locked.")).toOutcome(definition)
        assertSame(definition, outcome.definition)
    }

    @Test fun conversionRejectsNonFailureDefinitionDeterministically() {
        val pending =
            OutcomeDefinition.custom(
                "com.example.validation",
                "VALIDATION_PENDING",
                OutcomeState.PENDING,
                "Validation is pending.",
            )

        assertFails<IllegalArgumentException> {
            ValidationResult.valid().toOutcome(StandardOutcomes.OK)
        }
        assertFails<IllegalArgumentException> {
            ValidationResult.invalid(Issue.of("Bad.")).toOutcome(pending)
        }
    }

    @Test fun invalidHasValueEquality() {
        assertEquals(ValidationResult.invalid(Issue.of("Bad.")), ValidationResult.invalid(Issue.of("Bad.")))
        assertEquals(
            ValidationResult.invalid(Issue.of("Bad.")).hashCode(),
            ValidationResult.invalid(Issue.of("Bad.")).hashCode(),
        )
    }

    @Test fun invalidToStringContainsIssues() {
        assertTrue(ValidationResult.invalid(Issue.of("Bad.")).toString().contains("Bad."))
    }
}
